extension/encoding/awslogsencodingextension/unmarshaler/subscription-filter/subscription_filter.go

```diff
@@ -0,0 +1,140 @@
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	conventions "go.opentelemetry.io/collector/semconv/v1.27.0"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/awslogsencodingextension/internal/metadata"
+)
+
+var (
+	errEmptyOwner     = errors.New("cloudwatch log with message type 'DATA_MESSAGE' has empty owner field")
+	errEmptyLogGroup  = errors.New("cloudwatch log with message type 'DATA_MESSAGE' has empty log group field")
+	errEmptyLogStream = errors.New("cloudwatch log with message type 'DATA_MESSAGE' has empty log stream field")
+)
+
+func validateLog(log events.CloudwatchLogsData) error {
+	switch log.MessageType {
+	case "DATA_MESSAGE":
+		if log.Owner == "" {
+			return errEmptyOwner
+		}
+		if log.LogGroup == "" {
+			return errEmptyLogGroup
+		}
+		if log.LogStream == "" {
+			return errEmptyLogStream
+		}
+	case "CONTROL_MESSAGE":
+	default:
+		return fmt.Errorf("cloudwatch log has invalid message type %q", log.MessageType)
+	}
+	return nil
+}
+
+type subscriptionFilterUnmarshaler struct {
+	buildInfo component.BuildInfo
+
+	// We need to decompress the records.
+	// Creating a new gzip.Reader every time is expensive.
+	// We can keep the gzip in cache, and take advantage
+	// of the continuously AWS environment or the AWS
+	// warm starts, in case of lambda.
+	// See https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtime-environment.html.
+	gzipPool sync.Pool
+}
+
+func NewSubscriptionFilterUnmarshaler(buildInfo component.BuildInfo) plog.Unmarshaler {
+	return &subscriptionFilterUnmarshaler{
+		buildInfo: buildInfo,
+		gzipPool:  sync.Pool{},
+	}
+}
+
+var _ plog.Unmarshaler = (*subscriptionFilterUnmarshaler)(nil)
+
+// UnmarshalLogs deserializes the given record as CloudWatch Logs events
+// into a plog.Logs, grouping logs by owner (account ID), log group, and
+// log stream. Logs are assumed to be gzip-compressed as specified at
+// https://docs.aws.amazon.com/firehose/latest/dev/writing-with-cloudwatch-logs.html.
+func (f *subscriptionFilterUnmarshaler) UnmarshalLogs(compressedRecord []byte) (plog.Logs, error) {
+	var errDecompress error
+	record, ok := f.gzipPool.Get().(*gzip.Reader)
+	if !ok {
+		record, errDecompress = gzip.NewReader(bytes.NewReader(compressedRecord))
+	} else {
+		errDecompress = record.Reset(bytes.NewReader(compressedRecord))
+	}
+	if errDecompress != nil {
+		if record != nil {
+			f.gzipPool.Put(record)
+		}
+		return plog.Logs{}, fmt.Errorf("failed to decompress record: %w", errDecompress)
+	}
+	defer func() {
+		_ = record.Close()
+		f.gzipPool.Put(record)
+	}()
+
+	decompressedRecord, errRead := io.ReadAll(record)
+	if errRead != nil {
+		return plog.Logs{}, fmt.Errorf("failed to read decompressed record: %w", errRead)
+	}
+
+	var cwLog events.CloudwatchLogsData
+	if err := jsoniter.ConfigFastest.Unmarshal(decompressedRecord, &cwLog); err != nil {
```

Comment:
Can we pass `record` into a stream decoder, rather than reading all the uncompressed data into a buffer first?

Reply:
Done!